File: ai/app/avatar/parameter_mixer.py
# Parameter Mixer — resolves conflicts when multiple subsystems modify the same
# Cubism parameter simultaneously (e.g., lip sync + expression both write
# ParamMouthOpenY).
#
# Each subsystem registers its "owned" parameters with a priority.
# When multiple subsystems write to the same parameter, the mixer uses
# priority-weighted blending rather than last-writer-wins.
#
# Architecture:
#   LipSync ──┐
#   Blink   ──┤
#   Breath  ──┼── ParameterMixer ──→ Final parameter values
#   Express ──┤
#   Motion  ──┘

from dataclasses import dataclass, field
import logging
from typing import Any

logger = logging.getLogger("avatar.parameter_mixer")


@dataclass
class ParameterOwner:
    """Defines which subsystem owns a set of parameters."""
    name: str                # "lip_sync" | "blink" | "breath" | "expression"
    param_ids: list[str]     # ["ParamMouthOpenY"]
    priority: int            # 60, 40, 20, 30
    mask_weight: float = 1.0 # How strongly this owner contributes (0-1)


@dataclass
class ParameterValue:
    """A single parameter value from one source."""
    param_id: str            # "ParamMouthOpenY"
    value: float             # target value
    source: str              # "lip_sync" | "blink" | "breath" | "expression"
    priority: int            # source priority
    weight: float = 1.0      # blend weight for multi-source mixing


class ParameterMixer:
    """Resolves parameter conflicts using priority-weighted blending.

    Rules:
    - Each subsystem registers the parameters it "owns" with a priority.
    - If only one source writes to a parameter, its value is used directly.
    - If multiple sources write to the same parameter, values are blended
      using priority-weighted averaging.
    - A source with mask_weight=0 is effectively silent for its owned params.
    - blink has absolute priority on eye-open params when actively blinking
      (value < 0.5 means eyes are closing/closed).
    """

    def __init__(self):
        self._owners: dict[str, ParameterOwner] = {}
        # Current frame's parameter values: {param_id: [ParameterValue, ...]}
        self._frame_values: dict[str, list[ParameterValue]] = {}
        # Resolved output: {param_id: float}
        self._resolved: dict[str, float] = {}

    # ── Configuration ──────────────────────────────────────────────────

    def register_owner(self, name: str, param_ids: list[str],
                       priority: int, mask_weight: float = 1.0) -> None:
        """Register a subsystem as the owner of specific parameters."""
        self._owners[name] = ParameterOwner(
            name=name,
            param_ids=list(param_ids),
            priority=priority,
            mask_weight=mask_weight,
        )
        logger.debug("ParameterMixer: registered '%s' (priority=%d, params=%s)",
                     name, priority, param_ids)

    def register_all(self, config: dict[str, Any]) -> None:
        """Register multiple owners from avatar.yaml parameters section."""
        for name, cfg in config.items():
            self.register_owner(
                name=name,
                param_ids=cfg.get("owns", []),
                priority=cfg.get("priority", 10),
                mask_weight=cfg.get("mask_weight", 1.0),
            )

    def get_owner(self, name: str) -> ParameterOwner | None:
        """Get a registered owner by name."""
        return self._owners.get(name)

    def list_owners(self) -> list[str]:
        return list(self._owners.keys())

    # ── Per-frame value submission ─────────────────────────────────────

    def set_params(self, source: str, values: dict[str, float],
                   priority: int | None = None) -> None:
        """Submit parameter values from a source for the current frame.

        Args:
            source: Subsystem name ("lip_sync", "blink", etc.)
            values: {param_id: target_value} mapping
            priority: Override the source's registered priority.
                      If None, uses the registered owner priority.
        """
        owner = self._owners.get(source)
        src_priority = priority if priority is not None else (owner.priority if owner else 10)

        for param_id, value in values.items():
            pv = ParameterValue(
                param_id=param_id,
                value=value,
                source=source,
                priority=src_priority,
                weight=owner.mask_weight if owner else 1.0,
            )
            if param_id not in self._frame_values:
                self._frame_values[param_id] = []
            self._frame_values[param_id].append(pv)

    def reset_frame(self) -> None:
        """Clear all submitted values for the next frame."""
        self._frame_values.clear()

    # ── Resolution ─────────────────────────────────────────────────────

    def resolve(self) -> dict[str, float]:
        """Blend all submitted parameter values into final output.

        Returns {param_id: final_value}.
        """
        self._resolved = {}

        for param_id, values in self._frame_values.items():
            if len(values) == 1:
                # Single source — use directly
                self._resolved[param_id] = values[0].value
            else:
                self._resolved[param_id] = self._blend(param_id, values)

        return dict(self._resolved)

    def _blend(self, param_id: str, values: list[ParameterValue]) -> float:
        """Blend multiple values for the same parameter.

        Strategy:
        1. If blink is active (value < 0.5) on eye-open params, blink wins absolutely.
        2. Otherwise, priority-weighted average.
        3. Ties in priority: average the values.
        """
        # Absolute blink override: when eyes are closing/closed, blink wins.
        blink_values = [v for v in values if v.source == "blink"]
        if blink_values:
            # Blink active (eyes partially or fully closed)
            if any(v.value < 0.5 for v in blink_values):
                return blink_values[0].value  # Use first blink value

        # Priority-weighted blend
        total_weight = 0.0
        weighted_sum = 0.0

        for v in values:
            w = float(v.priority) * v.weight
            weighted_sum += v.value * w
            total_weight += w

        if total_weight == 0:
            return values[0].value  # fallback

        return weighted_sum / total_weight

    # ── Query ──────────────────────────────────────────────────────────

    def get_resolved(self, param_id: str) -> float | None:
        """Get the resolved value for a single parameter (after resolve())."""
        return self._resolved.get(param_id)

    def get_all_resolved(self) -> dict[str, float]:
        """Get all resolved values (after resolve())."""
        return dict(self._resolved)

    def debug_frame(self) -> dict:
        """Debug info: which sources contributed to which parameters."""
        return {
            "frame_values": {
                pid: [{"source": v.source, "value": v.value, "priority": v.priority}
                      for v in vals]
                for pid, vals in self._frame_values.items()
            },
            "resolved": dict(self._resolved),
        }

File: ai/frontend/src/character/ParameterMixer.ts
// Parameter Mixer — resolves conflicts when multiple subsystems write to the
// same Cubism parameter. This is the SINGLE arbitration point for all
// real-time parameter contributions.
//
// Flow:
//   1. Each subsystem submits ParameterContribution objects via submit() or setParams()
//   2. At resolve(), the mixer arbitrates per-parameter conflicts
//   3. apply() writes final values to the Live2DModelAdapter
//
// Priority conventions:
//   system safety/reset:  100
//   ui/debug override:     90
//   ai motion:             80
//   ai expression:         75
//   motion playback:       50
//   lip_sync:              60
//   pet interaction:       60
//   gaze:                  30
//   blink:                 40  (absolute override when eyes closing)
//   breath:                20
//   idle_sway:             10
//
// Channels:
//   motion, expression, lip_sync, eye, head, body, blink, breath, accessory, pose

import type { Live2DModelAdapter } from './Live2DModelAdapter'

export type ControlSource =
  | 'system'
  | 'runtime'
  | 'ai'
  | 'speech'
  | 'mouse'
  | 'idle'
  | 'pet'
  | 'ui'
  | 'debug'

export type ControlChannel =
  | 'expression'
  | 'lip_sync'
  | 'eye'
  | 'head'
  | 'body'
  | 'blink'
  | 'breath'
  | 'accessory'
  | 'pose'
  | 'calibration'
  | 'motion'

export interface ParameterContribution {
  /** Unique contribution ID for deduplication. */
  id: string
  /** Cubism parameter ID, e.g. "ParamAngleX". */
  parameterId: string
  /** Source identifier for debugging. */
  source: string
  /** Control channel for arbitration grouping. */
  channel: ControlChannel
  /** Parameter value. */
  value: number
  /** How this contribution combines with the resolved base value. */
  mode?: 'override' | 'add' | 'multiply'
  /** Priority (higher = wins conflicts). */
  priority: number
  /** Optional weight for blended contributions. */
  weight?: number
  /** Timestamp when this contribution was created. */
  createdAt: number
  /** Optional expiry — contribution is removed after this time. */
  expiresAt?: number
  /** Preserve this contribution across frames until explicitly removed. */
  persistent?: boolean
}

export interface PartOpacityContribution {
  id: string
  partId: string
  opacity: number
  priority: number
  weight?: number
  persistent?: boolean
}

interface ParameterOwner {
  name: string
  paramIds: string[]
  priority: number
  maskWeight: number
}

interface ParameterValue {
  paramId: string
  value: number
  source: string
  priority: number
  weight: number
  mode: 'override' | 'add' | 'multiply'
}

interface MixerDebugFrame {
  frameValues: Record<string, Array<{ source: string; value: number; priority: number }>>
  resolved: Record<string, number>
}

export class ParameterMixer {
  private _owners: Map<string, ParameterOwner> = new Map()
  private _frameValues: Map<string, ParameterValue[]> = new Map()
  private _resolved: Record<string, number> = {}
  // Persistent contributions (not cleared per-frame, e.g. accessories)
  private _persistentContributions: Map<string, ParameterContribution> = new Map()
  private _persistentPartOpacityContributions: Map<string, PartOpacityContribution> = new Map()
  private _partOpacityFrame: Map<string, PartOpacityContribution[]> = new Map()
  private _resolvedPartOpacities: Record<string, number> = {}
  private _baselineProvider: ((parameterId: string) => number) | null = null
  private _baselines = new Map<string, number>()

  setBaselineProvider(provider: ((parameterId: string) => number) | null): void {
    this._baselineProvider = provider
    this._baselines.clear()
  }

  private _baseline(parameterId: string): number {
    if (!this._baselines.has(parameterId)) {
      this._baselines.set(parameterId, this._baselineProvider?.(parameterId) ?? 0)
    }
    return this._baselines.get(parameterId)!
  }

  getBaseline(parameterId: string): number {
    return this._baseline(parameterId)
  }

  // ── Owner registration ─────────────────────────────────────

  registerOwner(name: string, paramIds: string[], priority: number, maskWeight = 1.0): void {
    this._owners.set(name, { name, paramIds, priority, maskWeight })
  }

  registerAll(config: Record<string, { owns: string[]; priority: number; mask_weight?: number }>): void {
    for (const [name, cfg] of Object.entries(config)) {
      this.registerOwner(name, cfg.owns, cfg.priority, cfg.mask_weight ?? 1.0)
    }
  }

  getOwner(name: string): ParameterOwner | undefined {
    return this._owners.get(name)
  }

  // ── Contribution API (NEW — preferred for all new code) ─────

  /**
   * Submit a parameter contribution for the current frame.
   * Contributions with the same id (for persistent ones) will be updated.
   */
  submit(contribution: ParameterContribution): void {
    // Only explicit persistent values or TTL values survive resetFrame().
    if (contribution.persistent || contribution.expiresAt) {
      this._persistentContributions.set(contribution.id, contribution)
    }
    this._addToFrame(contribution)
  }

  private _addToFrame(contribution: ParameterContribution): void {
    if (!this._frameValues.has(contribution.parameterId)) {
      this._frameValues.set(contribution.parameterId, [])
    }
    this._frameValues.get(contribution.parameterId)!.push({
      paramId: contribution.parameterId,
      value: contribution.value,
      source: contribution.source,
      priority: contribution.priority,
      weight: contribution.weight ?? 1.0,
      mode: contribution.mode ?? 'override',
    })
  }

  /**
   * Remove a persistent contribution by id.
   */
  removeContribution(id: string): void {
    this._persistentContributions.delete(id)
  }

  /** Queue a part opacity write so the adapter remains the only SDK writer. */
  submitPartOpacity(contribution: PartOpacityContribution): void {
    if (contribution.persistent) {
      this._persistentPartOpacityContributions.set(contribution.id, contribution)
    }
    this._addPartOpacityToFrame(contribution)
  }

  removePartOpacityContribution(id: string): void {
    this._persistentPartOpacityContributions.delete(id)
  }

  private _addPartOpacityToFrame(contribution: PartOpacityContribution): void {
    const values = this._partOpacityFrame.get(contribution.partId) ?? []
    const existingIndex = values.findIndex(value => value.id === contribution.id)
    if (existingIndex >= 0) values[existingIndex] = contribution
    else values.push(contribution)
    this._partOpacityFrame.set(contribution.partId, values)
  }

  // ── Legacy per-frame submission (for backward compat during migration) ──

  setParams(source: string, values: Record<string, number>, priority?: number): void {
    const owner = this._owners.get(source)
    const srcPriority = priority !== undefined ? priority : (owner?.priority ?? 10)
    const weight = owner?.maskWeight ?? 1.0

    for (const [paramId, value] of Object.entries(values)) {
      if (!this._frameValues.has(paramId)) {
        this._frameValues.set(paramId, [])
      }
      this._frameValues.get(paramId)!.push({
        paramId,
        value,
        source,
        priority: srcPriority,
        weight,
        mode: 'override',
      })
    }
  }

  resetFrame(now = performance.now()): void {
    this._frameValues.clear()
    this._partOpacityFrame.clear()
    for (const [id, contribution] of this._persistentContributions) {
      if (contribution.expiresAt !== undefined && contribution.expiresAt <= now) {
        this._persistentContributions.delete(id)
        continue
      }
      this._addToFrame(contribution)
    }
    for (const contribution of this._persistentPartOpacityContributions.values()) {
      this._addPartOpacityToFrame(contribution)
    }
  }

  // ── Resolution ──────────────────────────────────────────────

  /** Resolve all parameter conflicts. */
  resolve(): Record<string, number> {
    this._resolved = {}

    for (const [paramId, values] of this._frameValues.entries()) {
      this._resolved[paramId] = this._blend(paramId, values)
    }
    this.resolvePartOpacities()

    return { ...this._resolved }
  }

  resolvePartOpacities(): Record<string, number> {
    this._resolvedPartOpacities = {}
    for (const [partId, values] of this._partOpacityFrame) {
      const ordered = [...values].sort((a, b) => a.priority - b.priority)
      let resolved = 1
      for (const value of ordered) {
        const weight = Math.max(0, Math.min(1, value.weight ?? 1))
        const opacity = Math.max(0, Math.min(1, value.opacity))
        resolved += (opacity - resolved) * weight
      }
      this._resolvedPartOpacities[partId] = resolved
    }
    return { ...this._resolvedPartOpacities }
  }

  private _blend(paramId: string, values: ParameterValue[]): number {
    // Rule 1: Absolute blink override — when eyes are closing/closed, blink wins.
    const blinkValues = values.filter(v => v.source === 'blink')
    if (blinkValues.length > 0) {
      if (blinkValues.some(v => v.value < 0.5)) {
        return blinkValues[0].value
      }
    }

    // Exclusive arbitration for 'override' contributions: the highest-priority
    // override owns the parameter and fades from the model baseline by its own
    // weight. Lower-priority overrides never pre-load the value, so two
    // opposing poses cannot land the parameter in-between ("ghost arms").
    // 'add' and 'multiply' contributions accumulate on top of the resolved
    // override, matching the reference SDK's exclusive-group semantics.
    const overrides = values.filter(v => v.mode === 'override')
    const additions = values.filter(v => v.mode === 'add')
    const multipliers = values.filter(v => v.mode === 'multiply')

    let result = this._baseline(paramId)
    if (overrides.length > 0) {
      overrides.sort((a, b) => b.priority - a.priority || b.weight - a.weight)
      const winner = overrides[0]
      const weight = Math.max(0, Math.min(1, winner.weight))
      result += (winner.value - result) * weight
    }
    for (const contribution of additions) {
      result += contribution.value * Math.max(0, Math.min(1, contribution.weight))
    }
    for (const contribution of multipliers) {
      result *= 1 + (contribution.value - 1) * Math.max(0, Math.min(1, contribution.weight))
    }
    return result
  }

    /* Legacy unweighted arbitration retained only as migration reference.
    const overrides = values.filter(v => v.mode === 'override')
    const additions = values.filter(v => v.mode === 'add')
    const multipliers = values.filter(v => v.mode === 'multiply')

    // Rule 2: Highest priority wins for body/head/motion/arm parameters.
    //   Weighted average produces "ghost" positions (four arms) when two
    //   controllers write opposing override values — the arm appears in-between.
    if (overrides.length > 0) {
      overrides.sort((a, b) => b.priority - a.priority || b.weight - a.weight)
      let resolved = overrides[0].value
      for (const addition of additions) {
        resolved += addition.value * addition.weight
      }
      for (const multiplier of multipliers) {
        resolved *= multiplier.value
      }
      return resolved
    }

    // Pure additive/multiply: weighted average.
    let blended = 0
    let weightSum = 0
    for (const addition of additions) {
      const w = addition.priority * addition.weight
      blended += addition.value * w
      weightSum += w
    }
    let result = weightSum === 0 ? 0 : blended / weightSum
    for (const multiplier of multipliers) {
      result *= multiplier.value
    }
    return result
    */
  // ── Apply resolved values to adapter ─────────────────────────

  /** Apply resolved parameters to the Live2DModelAdapter (the ONLY write path). */
  apply(adapter: Live2DModelAdapter): void {
    for (const [paramId, value] of Object.entries(this._resolved)) {
      adapter.setParameter(paramId, value)
    }
    for (const [partId, opacity] of Object.entries(this._resolvedPartOpacities)) {
      adapter.setPartOpacity(partId, opacity)
    }
  }

  // ── Query ───────────────────────────────────────────────────

  getResolved(paramId: string): number | undefined {
    return this._resolved[paramId]
  }

  getAllResolved(): Record<string, number> {
    return { ...this._resolved }
  }

  getAllResolvedPartOpacities(): Record<string, number> {
    return { ...this._resolvedPartOpacities }
  }

  debugFrame(): MixerDebugFrame {
    const frameValues: Record<string, Array<{ source: string; value: number; priority: number }>> = {}
    for (const [pid, vals] of this._frameValues.entries()) {
      frameValues[pid] = vals.map(v => ({
        source: v.source,
        value: v.value,
        priority: v.priority,
      }))
    }
    return { frameValues, resolved: { ...this._resolved } }
  }
}
